cpp/unreal_plugins/SimulationController/Source/SimulationController/VehicleAgent.cpp

```diff
@@ -0,0 +1,283 @@
+        array_desc.shape_ = { 4 };
+        array_desc.datatype_ = DataType::Float64;
+        action_space["set_brake_torques"] = std::move(array_desc);
+    }
+
+    return action_space;
+}
+
+std::map<std::string, ArrayDesc> VehicleAgent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "location")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<double>::lowest();
+        array_desc.high_ = std::numeric_limits<double>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 3 };
+        observation_space["location"] = std::move(array_desc); // position (X, Y, Z) in [cms] of the agent relative to the world frame.
+    }
+
+    if (Std::contains(observation_components, "rotation")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<double>::lowest();
+        array_desc.high_ = std::numeric_limits<double>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 3 };
+        observation_space["rotation"] = std::move(array_desc); // orientation (Pitch, Yaw, Roll) in [degs] of the agent relative to the world frame.
+    }
+
+    if (Std::contains(observation_components, "wheel_encoder")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<double>::lowest();
+        array_desc.high_ = std::numeric_limits<double>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 4 };
+        observation_space["wheel_encoder"] = std::move(array_desc); // FL, FR, RL, RR, in [rad/s]
+    }
+
+    if (Std::contains(observation_components, "imu")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<double>::lowest();
+        array_desc.high_ = std::numeric_limits<double>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 6 };
+        observation_space["imu"] = std::move(array_desc); // a_x, a_y, a_z in [cm/s^s] g_x, g_y, g_z in [rad/s]
+    }
+
+    observation_space.merge(camera_sensor_->getObservationSpace(observation_components));
+
+    return observation_space;
+}
+
+std::map<std::string, ArrayDesc> VehicleAgent::getStepInfoSpace() const
+{
+    return {};
+}
+
+void VehicleAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& action)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "set_drive_torques")) {
+        vehicle_pawn_->setDriveTorques(Std::reinterpretAs<double>(action.at("set_drive_torques")));
+
+        // Apply the drive torque in [N.m] to the vehicle wheels.The applied drive torque persists until the
```

Comment:
Why commented?

Reply:
Since we fixed the issue with `Yield` macro, I've refined the code in this class.